### PR TITLE
SDK - Got rid of the global variable collecting all created pipelines

### PR DESCRIPTION
**Pull request by Ark-kun**
This list was only used by the command-line compiler.
The command-line compiler can still collect the created pipelines by registering a handler function in `_pipeline_decorator_handlers`.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/1167)

### CODE REVIEW

File: sdk/python/kfp/dsl/_pipeline.py [hunk 1]
```diff
@@ -19,6 +19,11 @@
 import sys
 
 
+# Contains a stack of handler functions.
```
Comment (gaoning777) [inline sdk/python/kfp/dsl/_pipeline.py]: Could you explain why we need the decorator handlers?

Comment (Ark-kun) [reply sdk/python/kfp/dsl/_pipeline.py]: This is a bit similar to the "dependency injection" design pattern.

Right know, the `@pipeline` decorator is always running some code that is only needed in a specific compiler scenario. The decorator has a dependency on that compiler-related code.
I'm removing that dependency.

But during some command-line compilation scenarios the code still needs to be run. So I'm adding a way for the command-line compiler to add a handler and run the code it wants (but only when it's needed, not every time).
That's how we enable the `@pipeline` to run the command-line compiler specific code without introducing a dependency.

Comment (gaoning777) [reply sdk/python/kfp/dsl/_pipeline.py]: Got it. Thanks for the explanation.

File: sdk/python/kfp/dsl/_pipeline.py [hunk 2]
```diff
@@ -19,6 +19,11 @@
...
+# The last handler in the list is called whenever the @pipeline decorator is used
+_pipeline_decorator_handlers = []
```
Comment (gaoning777) [inline sdk/python/kfp/dsl/_pipeline.py]: Maybe we only need one decorator handler? What are the scenarios where multiple decorator handlers are needed?

File: sdk/python/kfp/dsl/_pipeline.py (hunk 1 above)
Comment (gaoning777) [inline sdk/python/kfp/dsl/_pipeline.py]: Could we add more comments in the codes to explain the decorator handlers.
To my understanding, it is a placeholder in the SDK such that the compiler could inject the any codes it wants to run.

File: sdk/python/kfp/dsl/_pipeline.py (hunk 2 above)
Comment (Ark-kun) [reply sdk/python/kfp/dsl/_pipeline.py]: We only call the last one, so we only need one.
But we need to put the old one somewhere.
Having a stack makes it easier as you do not need to remember the old handler - you just call `.pop()`.

If you want, I can replace this handler stack with a single handler.

File: sdk/python/kfp/dsl/_pipeline.py [hunk 3]
```diff
@@ -82,31 +90,16 @@ class Pipeline():
...
-    """Get all imported pipeline functions (decorated with @pipeline)."""
-    return Pipeline._pipeline_functions
-
-  @staticmethod
-  def _add_pipeline_to_global_list(func):
-    """Add a pipeline function (decorated with @pipeline)."""
-    Pipeline._pipeline_functions.append(func)
-
   @staticmethod
   def add_pipeline(name, description, func):
```
Comment (gaoning777) [inline sdk/python/kfp/dsl/_pipeline.py]: maybe we could remove the add_pipeline function here if we instruct the users to only use the decorators.

File: sdk/python/kfp/dsl/_pipeline.py (hunk 1 above)
Comment (Ark-kun) [reply sdk/python/kfp/dsl/_pipeline.py]: Sure. I can add a comment explaining it in more details.

File: sdk/python/kfp/dsl/_pipeline.py (hunk 3 above)
Comment (Ark-kun) [reply sdk/python/kfp/dsl/_pipeline.py]: Well... we did instruct the users.
In 3+ places.
We even raised exceptions in the compiler.
Unfortunately, Ajay has decided to use this function in TFX code. See https://github.com/tensorflow/tfx/pull/39 for details.
I've prepared the fix, but now we cannot just delete this function - we have to deprecate it over time =(

Comment (gaoning777) [reply sdk/python/kfp/dsl/_pipeline.py]: Make sense. If the TFX sample is using this function, let's keep it for now.

Comment (neuromage) [reply sdk/python/kfp/dsl/_pipeline.py]: Decorators are great for users when constructing a pipeline. However, decorators are syntactic sugar, and ideally, we should also provide a public-facing API to achieve the same result.  Consider a power-user, who wants to build 100 different pipelines, each with slightly different parameters/components within it. By removing this method, we make it not possible for the user to do this. I'm all for cleaning the code up and not exposing internals, but I think we should offer a public API for users to programmatically build pipelines. Right now, we seem to be removing public (by definition) API methods, which is a little strange to me.

What do you think?
/cc @paveldournov 
/cc @vicaire 
/cc @IronPan

Comment (Ark-kun) [reply sdk/python/kfp/dsl/_pipeline.py]: >Consider a power-user, who wants to build 100 different pipelines, each with slightly different parameters/components within it.
>By removing this method, we make it not possible for the user to do this.

That's not the case. Why do you think so?
You can have hundreds of pipelines. You do not need that undocumented function for that.

You can apply the decorator to 100s of pipelines.
Also, decorator is just a function. You can call a function on 100s of pipelines.
I see no problem.
And I see no reason why someone would want to circumvent it.
There might be some misunderstanding on how the decorators and DSL compiler work. I'll be happy to clear those misunderstandings.

Comment (Ark-kun) [reply sdk/python/kfp/dsl/_pipeline.py]: Also, as you see, I'm not removing this method now. That's unfortunate, but that TFX code was released and we now have to support it for a while. I wish our team could review it before the release.

File: sdk/python/kfp/dsl/_pipeline.py (hunk 2 above)
Comment (Ark-kun) [reply sdk/python/kfp/dsl/_pipeline.py]: I've switched to single handler.

File: sdk/python/kfp/dsl/_pipeline.py (hunk 1 above)
Comment (Ark-kun) [reply sdk/python/kfp/dsl/_pipeline.py]: Added a better description.

**Pull request merged**